### PR TITLE
Update assay upload statuses in preparation for cloud function merge status tracking

### PR DESCRIPTION
**Pull request by jacoblurye**
This PR should not be merged until:
* [x] an `/ingestion/poll_upload_merge_status` endpoint is added that instructs the CLI how long to wait before retrying upload status checks: **see `ingestion/poll_upload_merge_status`**
* [x] the CLI has been updated to accommodate these changes - merging the PR will cause the current CLI version's uploads to fail with 400s on staging, because the CLI uses old upload status values ("errored" and "completed"): **https://github.com/CIMAC-CIDC/cidc-cli/pull/96**
* [x] the `ingest_upload` cloud function has been updated to set an assay upload's status on merge success or failure: **https://github.com/CIMAC-CIDC/cidc-cloud-functions/pull/36**

### CODE REVIEW

**Comment by jim-bo** [comment]
@jacoblurye I'll leave detailed comments to @curlup but the concept looks good to me.

File: cidc_api/services/ingestion.py [hunk 1]
```diff
@@ -293,6 +299,52 @@ def upload_assay():
...
+        raise BadRequest("Missing expected query parameter 'id'")
+
+    upload = AssayUploads.find_by_id(upload_id)
+
+    if not upload:
+        raise NotFound(f"Could not find assay upload job with id {upload_id}")
+
+    # Users should only be able to poll the status of their own uploads
+    user = _request_ctx_stack.top.current_user
+    if not user.email == upload.uploader_email:
```
Comment (curlup) [inline cidc_api/services/ingestion.py]: shouldn't this go as `AssayUploads.find_by_id.filter()`

File: migrations/versions/571d8a2570a6_.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    "merge-completed",
+    "merge-failed",
+    name="assay_upload_status",
+)
+
+
+def upgrade():
+    assay_upload_status.create(op.get_bind())
+
+    op.execute("ALTER TABLE assay_uploads ALTER COLUMN status TYPE text")
```
Comment (curlup) [inline migrations/versions/571d8a2570a6_.py]: From the model def I thought this column will be renamed. Shouldn't it be?

Comment (jacoblurye) [reply migrations/versions/571d8a2570a6_.py]: I don't think so. `AssayUploads` still have a `status` field; this PR adds an additional `AssayUploads.status_details` field in the other migration (below).

File: cidc_api/gcloud_client.py [hunk 3]
```diff
@@ -131,6 +131,11 @@ def _encode_and_publish(content: str, topic: str) -> Future:
     topic = pubsub_publisher.topic_path(GOOGLE_CLOUD_PROJECT, topic)
     data = bytes(content, "utf-8")
 
+    # Don't actually publish to Pub/Sub if running locally
+    if ENV == "dev":
+        print(f"Would publish message {content} to topic {topic}")
```
[the remark this answers is not in the document]
Comment (jacoblurye) [reply cidc_api/gcloud_client.py]: There's also the official pub/sub emulator (just found the link, haven't use this before): https://cloud.google.com/pubsub/docs/emulator

File: cidc_api/models.py [hunk 4]
```diff
@@ -519,6 +551,14 @@ def create(
 
         return job
 
+    @classmethod
+    @with_default_session
+    def find_by_id(cls, id, email, session):
```
Comment (curlup) [inline cidc_api/models.py]:
```suggestion
    def find_by_id_and_email(cls, id, email, session):
```

File: cidc_api/services/ingestion.py [hunk 5]
```diff
@@ -293,6 +300,71 @@ def upload_assay():
...
+        400: no "id" query parameter is supplied
+        401: the requesting user did not create the requested upload job
+        404: no upload job with id "id" is found
+    """
+    upload_id = request.args.get("id")
+    if not upload_id:
+        raise BadRequest("Missing expected query parameter 'id'")
+
+    user = _request_ctx_stack.top.current_user
+    upload = AssayUploads.find_by_id(upload_id, user.email)
```
Comment (curlup) [inline cidc_api/services/ingestion.py]:
```suggestion
    upload = AssayUploads.find_by_id_and_email(upload_id, user.email)
```

**Pull request merged**